proposals/NNNN-global-actor-isolated-types-usability.md

```diff
@@ -0,0 +1,267 @@
+
+```swift
+class NotSendable {}
+
+
+@MainActor
+class Subclass: NotSendable {} // error: main actor-isolated class 'Subclass' has different actor isolation from nonisolated superclass 'NotSendable'
+```
+
+Because global actor isolation on a class implies a `Sendable` conformance, adding isolation to a subclass of a non-`Sendable` superclass can circumvent `Sendable` checking:
+
+```swift
+func computeCount() async -> Int { ... }
+
+class NotSendable {
+  var mutableState = 0
+  func mutate() async {
+    let count = await computeCount()
+    mutableState += count
+  }
+}
+
+@MainActor
+class Subclass: NotSendable {}
+
+func test() async {
+  let c = Subclass()
+  await withDiscardingTaskGroup { group in
+    group.addTask {
+      await c.mutate()
+    }
+
+    group.addTask { @MainActor in
+      await c.mutate()
+    }
+  }
+}
+```
+
+In the above code, an instance of `Subclass` can be passed across isolation boundaries because `@MainActor` implies that the type is `Sendable`. However, `Subclass` inherits non-isolated, mutable state from the superclass, so this `Sendable` conformance allows smuggling unprotected shared mutable state across isolation boundaries to create potential for concurrent access. For this reason, the warning about adding isolation to a subclass was added in Swift 5.10, but this restriction could be lifted by instead preventing the subclass from being `Sendable`.
+
+## Proposed solution
+
+We propose that:
+
+- `Sendable` properties of a global-actor-isolated value type would be treated `nonisolated` as inferred within the module.
+- `@Sendable` would be inferred for global-actor-isolated functions and closures. Additionally, globally isolated closures would be allowed to capture non-`Sendable` values.
+- The programmer would be able to suppress the automatic conformance inferred via the above rule using the new `@~Sendable` attribute. By analogy, introduce a new `~Sendable` protocol to indicate that a nominal type is not `Sendable`.
+- Require the global-actor-isolated subclass of a `nonisolated`, non-`Sendable` to be non-`Sendable`.
+
+
+## Detailed design
+
+
+### Inference of `nonisolated` for `var` properties of globally isolated value types
+
+Let's look at the first problem with usability of a `var` property of a main-actor-isolated struct:
+
+```swift
+@MainActor 
+struct S {
+  var x: Int = 0 // okay ('nonisolated' is inferred within the module)
+}
+
+extension S: Equatable {
+  static nonisolated func ==(lhs: S, rhs: S) -> Bool {
+    return lhs.x == rhs.x // okay
+  }
+}
+```
+
+In the above code, `x` is implicitly `nonisolated` within the module. Under this proposal, `nonisolated` is inferred for within the module access of `Sendable` properties of a global-actor-isolated value type. This is data-race safe because the property belongs to a value type, meaning it will be copied every time it crosses an isolation boundary.
```

Comment:
You don't need to re-justify things here that you already justified in the earlier sections, and you can let the section title introduce the section.

This paragraph is a little confusing.  You need to be clear that you're not talking about actually inferring `nonisolated` — you're saying that certain accesses will be allowed *as if* the property were declared `nonisolated`.  You should say that this rules only applies to stored properties, and you may need to discuss the impact on property wrappers and `lazy`.  You should mention that the same effect happens across modules if the type is `@frozen` (right?).

Please use both positive and negative examples of things that are now allowed and still not allowed.  I think you should include an example of a mutation of a property.

Reply:
Thank you!! We are working on rewriting this paragraph.

I have some questions about the negative example: 

The mutation of a property is still not allowed, because the only way to do so is to use `inout`. As a consequence, the property becomes non-`Sendable` because it is now wrapped in `LValueType`, so our change has no effect. I don't think it's a good "negative" example just because of how `inout` works.

I could include an example where the mutated version of the value type is assigned into a new value and returned, but it would make it obvious that it was just copied. So I'm not sure whether such an example would be worth to include.

Reply:
As I understand it, mutation should be a positive example.  This code should be allowed under the proposal:

```swift
@MainActor struct A {
  var x: Int
}

func foo() {
  var a = A(x: 5)
  a.x += 5
}
```

Currently it is not:

```text
red.swift:7:5: error: main actor-isolated property 'x' can not be mutated from a non-isolated context
1 │ @MainActor struct A {
2 │   var x: Int
  │       ╰─ note: mutation of this property is only permitted within the actor
3 │ }
4 │ 
5 │ func foo() {
  │      ╰─ note: add '@MainActor' to make global function 'foo()' part of global actor 'MainActor'
6 │   var a = A(x: 5)
7 │   a.x += 5
  │     ╰─ error: main actor-isolated property 'x' can not be mutated from a non-isolated context
8 │ }
```

Reply:
Under this proposal, this code still gives me an error
```
119 │ func foo() {
120 │   var a = A(x: 5)
121 │   a.x += 5
    │   ╰─ warning: non-sendable type '@lvalue A' passed in call to main actor-isolated property 'x' cannot cross actor boundary; this is an error in the Swift 6 language mode
122 │ }
123 │ 
```
This is because Sendable can't be inferred from LValue type

Reply:
That certainly seems like something we should fix.

Reply:
Yeah, make LValue types conform to Sendable?

Reply:
The analysis probably needs to think more carefully about `inout` parameters, but yes, something like that.